### PR TITLE
feat(codegen): compile multiple L2 Orchestrations in one program

### PR DESCRIPTION
**Pull request by lyfne123**
## Summary

A PyPTO program can already define more than one L2 ``Orchestration``
function, but only the first one used to be compiled — ``_generate_single_chip``
silently picked the head orch via ``next(...)`` and dropped the rest. This PR
makes every L2 orch in a program compile and become independently callable
without requiring an L3 HOST wrapper.

### Codegen

Top-level ``generate()`` now detects ``orch_count > 1`` and dispatches to a
new ``_generate_multi_chip``, which emits each Orchestration as a
self-contained sub-build under ``next_levels/{orch_name}/`` — the same
layout ``_generate_with_distributed`` already produces for the chip tier
of L3+ programs. ``host_orch.py`` / ``sub_workers/`` are intentionally
not emitted because there is no L3 host driver.

### User-facing API

``CompiledProgram`` grows a sub-dispatch surface so callers can choose
which orch to run:

```python
compiled = ir.compile(MyTwoOrchProgram)
compiled.orchestration_names           # ['chip_orch_add', 'chip_orch_sub']
compiled.chip_orch_add(a, b, f_add)    # attribute lookup
compiled['chip_orch_sub'](a, b, f_sub) # subscript lookup
compiled(a, b, f)                      # TypeError — asks user to pick
```

Single-orch programs are unchanged; ``compiled(...)`` still works as
before.

### Refactor

``_extract_param_infos``, ``_build_full_args``, ``_validate_device_tensor``
and the per-call dispatch body are extracted into module-level helpers
(``_extract_func_param_infos`` / ``_invoke_compiled``) so the single-orch
``CompiledProgram.__call__`` and the new multi-orch ``_SubChipCallable.__call__``
share one path.

## Testing

- ``tests/ut/ir/test_compiled_program.py``: updated import to use the
  newly module-level ``_build_full_args``. All existing single-orch unit
  tests continue to pass (38/38).
- ``tests/st/distributed/test_l2_multi_orch.py`` *(new)*:
  - ``test_codegen_emits_per_orch_subdirs`` — verifies both orchs land
    in ``next_levels/<name>/`` with the right kernels and that no flat
    ``orchestration/`` directory is left at the root.
  - ``test_dispatch_surface`` — covers ``compiled.orchestration_names``,
    subscript / attribute access, and the error paths
    (``compiled(...)`` rejection, ``KeyError`` for unknown orch,
    ``AttributeError`` for unknown attribute that doesn't shadow an orch).
  - ``test_execute_both_orchs_on_device`` — gated on ``--device`` and
    ``not --codegen-only``; compiles with ptoas and dispatches both
    orchs through subscript and attribute access, asserting their
    outputs are independent and correct.

## Test plan

- [x] Codegen + dispatch surface tests pass under ``--codegen-only`` on macOS
- [x] Full unit-test sweep on ``tests/ut/ir/test_compiled_program.py`` (38 passing)
- [x] Rebased onto upstream main; adopts new ``execute_compiled(dfx=..., block_dim=..., aicpu_thread_num=...)`` signature
- [ ] On-device execution of ``test_execute_both_orchs_on_device`` on a host with Ascend toolchain + ``--device``

### CODE REVIEW

File: python/pypto/ir/compiled_program.py [hunk 1]
```diff
@@ -105,6 +105,41 @@ class _ParamInfo:
     dtype: DataType
 
 
+def _extract_func_param_infos(func: Any) -> tuple[list[_ParamInfo], list[int], list[Any]]:
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/ir/compiled_program.py]: Fixed in b108cf33 — typed as `Function`.

File: python/pypto/ir/compiled_program.py [hunk 2]
```diff
@@ -318,139 +516,73 @@ def __call__(  # noqa: PLR0912 — dispatch logic with several arg-shape branche
...
+    __test__ = False
 
-        if not return_style:
-            return None
-        # Output indices always correspond to tensor params (scalars cannot be Out
-        # direction), so the elements are guaranteed to be torch.Tensor.
-        outputs = [coerced[i] for i in output_indices]
-        assert all(isinstance(o, torch.Tensor) for o in outputs)
-        return outputs[0] if len(outputs) == 1 else tuple(outputs)  # type: ignore[return-value]
+    def __init__(self, name: str, func: Any, sub_dir: Path, platform: str) -> None:
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/ir/compiled_program.py]: Fixed in b108cf33 — typed as `Function`.

File: python/pypto/ir/compiled_program.py [hunk 3]
```diff
@@ -133,33 +169,147 @@ def _extract_param_infos(program: Program) -> tuple[list[_ParamInfo], list[int],
...
             raise TypeError(
-                f"Unsupported parameter type for {param.name_hint!r}: {type(param_type).__name__}. "
-                f"Expected ShapedType or ScalarType."
+                f"Parameter {info.name!r} expects shape {expected_shape}; got DeviceTensor shape {arg.shape}"
             )
+    expected_dtype = _to_torch_dtype(info.dtype)
+    if expected_dtype is not None and arg.dtype != expected_dtype:
+        raise TypeError(
+            f"Parameter {info.name!r} expects dtype {expected_dtype}; got DeviceTensor dtype {arg.dtype}"
+        )
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/ir/compiled_program.py:203]: Fixed in `1beba4ee` — now enforces rank and checks each static dim individually, so partially-dynamic signatures like `[128, -1]` reject a wrong leading dim. Dropped the all-static gate.

File: python/pypto/ir/compiled_program.py [hunk 4]
```diff
@@ -219,7 +369,28 @@ def __init__(
...
+        if next_levels.is_dir():
+            for child in sorted(next_levels.iterdir()):
+                if child.is_dir() and (child / "orchestration").is_dir():
+                    self._sub_chip_dirs[child.name] = child
+
+        # Debug runner only makes sense when there's a single canonical entry
+        # point; multi-orch programs have one sub-build (and one debug script)
+        # per orch, emitted by each sub-build's own pipeline.
+        if not self._sub_chip_dirs:
+            self._emit_debug_runner()
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/ir/compiled_program.py:409]: Fixed in `1beba4ee` — skip multi-orch detection when `orchestration/host_orch.py` is present at the root (L3+ distributed builds). Distributed `CompiledProgram`s now keep `_sub_chip_dirs` empty and continue to go through `__call__`.

File: tests/st/distributed/test_l2_multi_orch.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        assert add_via_sub.param_names == ["a", "b", "f"]
+        assert sub_via_sub.param_names == ["a", "b", "f"]
+
+        # 4) Plain compiled(...) refuses — user must select an orch explicitly
+        a = torch.zeros((128, 128), dtype=torch.float32)
+        with pytest.raises(TypeError, match="select one explicitly"):
+            compiled(a, a, a)
+
+        # 5) Unknown orch name raises KeyError listing what *is* available
+        with pytest.raises(KeyError, match="chip_orch_add.*chip_orch_sub"):
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply tests/st/distributed/test_l2_multi_orch.py]: Fixed in `1beba4ee` — switched to `r"chip_orch_add.*chip_orch_sub"`.

**Review by Hzfengsy — APPROVED** [review]
